cpp/src/arrow/compute/kernels/vector_sort.cc

```diff
@@ -190,6 +236,67 @@ class ChunkedArraySorter : public TypeVisitor {
   NullPartitionResult* output_;
 };
 
+template <>
+void ChunkedArraySorter::MergeNonNulls<DictionaryType>(
+    uint64_t* range_begin, uint64_t* range_middle, uint64_t* range_end,
+    const std::vector<const Array*>& arrays, uint64_t* temp_indices) {
+  const ChunkedArrayResolver left_resolver(arrays);
+  const ChunkedArrayResolver right_resolver(arrays);
+
+  // concatenate all dictionary arrays to calculate rank
+  ArrayVector dicts_array;
+  for (const auto& array : arrays) {
+    const auto& dicts = checked_cast<const DictionaryArray&>(*array);
+    dicts_array.push_back(dicts.dictionary());
+  }
+  auto concat_dict = Concatenate(dicts_array).ValueOrDie();
+  auto ranks = RanksWithNulls(concat_dict).ValueOrDie();
+
+  // build unified rank map using dicts and rank array
+  std::unordered_map<std::string, int64_t> unified_rank_map;
+  for (int i = 0; i < concat_dict->length(); i++) {
+    auto dict = concat_dict->GetScalar(i).ValueOrDie();
```

Comment:
We tend to avoid GetScalar because it's expensive (It unwraps the array as its real type, gets the value, and wraps it as the scalar class again). Instead we often do `checked_cast<ConcreteArrayType&>(*arr)->GetView(i)` to directly get its ctype value.